
project vendor/opengapps/build/
diff --git a/opengapps-packages.mk b/opengapps-packages.mk
index 407a407..f03c0b1 100644
--- a/opengapps-packages.mk
+++ b/opengapps-packages.mk
@@ -25,9 +25,15 @@ GAPPS_PRODUCT_PACKAGES += \
     GoogleServicesFramework \
     GoogleLoginService \
     SetupWizard \
-    Phonesky \
     GoogleCalendarSyncAdapter
 
+ifeq ($(BOARD_IS_AUTOMOTIVE), true)
+GAPPS_PRODUCT_PACKAGES += PhoneskyCar 
+
+else
+GAPPS_PRODUCT_PACKAGES += Phonesky 
+endif
+
 TVGAPPS_PRODUCT_PACKAGES += \
     ConfigUpdater \
     GoogleBackupTransport \


